consult-yasnippet.el

```diff
@@ -0,0 +1,140 @@
+
+(defun consult-yasnippet--preview ()
+  "Previewer for `consult--read'.
+This function expands TEMPLATE at point in the buffer
+`consult-yasnippet--read-snippet' was started in. This includes
+overwriting any region that was active and removing any previous
+previews that're already active.
+
+When TEMPLATE is not given, this function essentially just resets
+the state of the current buffer to before any snippets were previewed."
+  (let* ((buf (current-buffer))
+         (region (if (region-active-p)
+                     (cons (region-beginning) (region-end))
+                   (cons (point) (point))))
+         (region-contents (buffer-substring (car region) (cdr region))))
+    (lambda (template restore)
+      (with-current-buffer buf
+        (let ((yas-verbosity 0)
+              (inhibit-redisplay t)
+              (inhibit-read-only t)
+              (orig-offset (- (point-max) (cdr region)))
+              (yas-prompt-functions '(yas-no-prompt)))
+
+          ;; We always undo any snippet previews before maybe setting up
+          ;; some new previews.
+          (delete-region (car region) (cdr region))
+          (goto-char (car region))
+          (setcar region (point))
+          (insert region-contents)
+          (setcdr region (point))
+
+          (when (and template (not restore))
+            (unwind-protect
+                (consult-yasnippet--expand-template template region region-contents)
+              (unwind-protect
+                  (mapc #'yas--commit-snippet
+                        (yas-active-snippets (point-min) (point-max)))
+                (setcdr region (- (point-max) orig-offset))))
+            (redisplay)))))))
+
+(defun consult-yasnippet--candidates (&rest body)
+  "Retrieve the list of available snippets in the current buffer."
+  (unless (bound-and-true-p yas-minor-mode)
+    (error "`yas-minor-mode' not enabled in current buffer"))
+
+  (cl-loop for template in (yas--all-templates (yas--get-snippet-tables))
+           with annotation = nil
+           ;; TODO: custom face
+           do (setq annotation
+                    (propertize (yas--template-key template) 'face 'font-lock-type-face))
+           collect (cons (concat (yas--template-name template)
+                                     (propertize " " 'display (concat " [" annotation "]")))
+                         template)))
+
+(defun consult-yasnippet--read-snippet ()
+  "Backend implementation of `consult-yasnippet'.
+
+This starts a `completing-read' session with all the snippets in the current
+snippet table with support for previewing the snippet to be expanded and
+replacing the active region with the snippet expansion.
+
+This function doesn't actually expand the snippet, it only reads and then
+returns a snippet template from the user."
+  (barf-if-buffer-read-only)
+
+  (let* ((buffer-undo-list t))                                                  ; Prevent querying user (and showing previews) from updating the undo-history
```

Comment:
Since you disable undo here, it won't work for the final expansion of the snippet. It should be possible to undo this.

Reply:
I'm not having any issues using evil-undo and undo-tree. And I don't see why there would be any issue. The final expansion doesn't happen in `consult-yasnippet--read-template`, this command just presents snippets as completion candidates and returns the chosen snippet. The final expansion happens at `consult-yasnippet`.

Reply:
Ah okay, let me check this. I may have missed that.

Reply:
You are right. This should be good then!